src/libsyntax/parse/attr.rs

```diff
@@ -55,26 +76,49 @@ impl<'a> Parser<'a> {
     /// If permit_inner is true, then a leading `!` indicates an inner
     /// attribute
     pub fn parse_attribute(&mut self, permit_inner: bool) -> PResult<'a, ast::Attribute> {
-        debug!("parse_attributes: permit_inner={:?} self.token={:?}",
+        debug!("parse_attribute: permit_inner={:?} self.token={:?}",
                permit_inner,
                self.token);
+        let inner_parse_policy = if permit_inner {
+            InnerAttributeParsePolicy::Permitted
+        } else {
+            InnerAttributeParsePolicy::NotPermitted
+                { reason: DEFAULT_UNEXPECTED_INNER_ATTR_ERR_MSG }
+        };
+        self.parse_attribute_with_inner_parse_policy(inner_parse_policy)
+    }
+
+    /// The same as `parse_attribute`, except it takes in an `InnerAttributeParsePolicy`
+    /// that prescribes how to handle inner attributes.
+    fn parse_attribute_with_inner_parse_policy(&mut self,
+                                               inner_parse_policy: InnerAttributeParsePolicy)
+                                               -> PResult<'a, ast::Attribute> {
+        debug!("parse_attribute_with_inner_parse_policy: inner_parse_policy={:?} self.token={:?}",
+               inner_parse_policy,
+               self.token);
         let (span, value, mut style) = match self.token {
             token::Pound => {
                 let lo = self.span.lo;
                 self.bump();
 
-                if permit_inner {
+                if inner_parse_policy == InnerAttributeParsePolicy::Permitted {
                     self.expected_tokens.push(TokenType::Token(token::Not));
                 }
                 let style = if self.token == token::Not {
                     self.bump();
-                    if !permit_inner {
+                    if let InnerAttributeParsePolicy::NotPermitted { reason } = inner_parse_policy
+                    {
                         let span = self.span;
                         self.diagnostic()
-                            .struct_span_err(span,
-                                             "an inner attribute is not permitted in this context")
+                            .struct_span_err(span, reason)
                             .help("place inner attribute at the top of the module or \
                                    block")
+                            .note("an inner attribute is one of the form `#![foo]` in \
+                                   `mod bar { #![foo] }`, modifying the item enclosing it \
+                                   (`mod` bar in our example)")
+                            .note("an outer attribute is either of the form `#[foo]` in \
+                                   `#[foo] struct Baz` or is an outer doc comment, modifying \
+                                   the item following it (`struct Baz` in our example)")
```

Comment:
This is a lot of spew. Let's try to condense a bit. First, let's remove the 'help'. I think with the new explanation it is pretty useless.

For the 'note' it would be good to indicate that inner attributes can be comments and are usually encountered at the crate level.

What about this: "inner attributes and doc comments, like `#![no_std]` or `//! My crate`, annotate the enclosing item, and are usually found at the crate root. Outer attributes and doc comments, like `#[test]` and `/// My function`, annotate the following item."